### ISSUE TITLE
Migration from Paperclip to ActiveStorage?

### ISSUE THREAD

**Issue by shuheiktgw** [tootsuite/mastodon#7414]
Paperclip maintainers recommends a migration to ActiveStorage.
We need to tackle this problem in the future?

ref: https://github.com/thoughtbot/paperclip/pull/2568

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by nightpool** [comment]
Oh, looks like you have to checksum *all* of the existing data. that's gonna be rough, especially on installs with hundreds of gigs of media

**Comment by nightpool** [comment]
ugh, and we're going to have to worry about n+1 problems *everywherE*

**Comment by Gargron** [comment]
Right, it's just going to take forever for any large server, and we're going to deal with n+1 query issues because data that used to be embedded within a single record is now going to be in a separate table. Tbh I'm not sure I'm a fan of that design except for the deduplication possibility.

**Comment by ThisIsMissEm** [comment]
Well, surely it won't be n+1 queries as long as we correctly use joins & includes where appropriate?

**Comment by Gargron** [comment]
Right, it's just a lot of places where it has to be changed, potentially, and extra database queries.

**Comment by TheEssem** [comment]
Is this still under consideration, by any chance? It would allow for using libvips instead of ImageMagick for image processing which would offer massive performance and security improvements.

**Comment by renchap** [comment]
This migration is probably a good idea, but will be a huge undertaking in both:
- the needed code changes
- the migration on every instances

I am not (yet) very familiar with Mastodon's source code, but from what I saw it looks like to be a quite "normal" Rails codebase. I am used to working with a huge number of media files (I run [notos.co](notos.co) with dozen of million of files, using Rails & [Shrine](https://shrinerb.com)).

I see a few possible paths here:

### 1 — Not migrating to a new storage.

This is probably the easiest way (nothing to do!) but PaperClip is no longer maintained


### 2 — Migrating to a new storage in 2 steps

Implement the new storage system in parallel of the existing one. New model methods are created (they can have the same interface as the current ones if needed) and will either
1. use the new storage if there is an attachment returned by ActiveStorage
2. use the old storage if there is no data in ActiveStorage

When a new file is attached, is it stored in ActiveStorage.

Once this is done, we can release a new version and ask instance admins to deploy it and their files will start to be stored into ActiveStorage's backend.

Then we provide a task to migrate existing PaperClip attachments to ActiveStorage. If this task is ran a few weeks after the first one, most files will probably have moved their location already. It should probably not directly do the work, but enqueue jobs in Sidekiq for file migrations (with a low priority queue?).

Finally, in the next major version we drop the old PaperClip fields and code. There should probably be a check here to ensure the is no data left in any PaperClip fields so instances do not loose any data.

This way of doing is much more complex and will require a lot of documentation about how to upgrade and how to switch to the new system.

### 3 — Migrating to a new storage, in one go

ie, during the Mastodon instance upgrade to a new version (using an AR migration)

This is probably the easiest way to migrate for instance admins, but is not doable if the files themselves need to be moved or any additional data needs to be computed during the migration (for example the checksum).

This is not doable with ActiveStorage at all (checksums needs to be copied, files needs to be moved).

It is definitely doable using [Shrine](https://shrinerb.com), as Shrine stores the full path the the attachment into it's JSON field, so the existing path can be kept for old attachments and migration is "only" a matter of converting the existing fields to Shrine's JSON field. Metadata for existing files (in Shrine's terms, file size, media dimensions, blurhash, checksum) can be left out in the migration, and be re-computed later (or only added for new attachments).

A [Paperclip migration guide exists](https://shrinerb.com/docs/paperclip) and shows a possible way of doing so.

Shrine has a lot of nice plugins and is very nice to work with, **this is the solution I would personally recommend**. It also has the advantage that you can use multiple storages across your application, allowing instance owners to split the various files (local accounts / remote accounts, splitting persistent storage and transient storage like cached files…).

It also allows migration from one storage to another one, or easily switching from one storage to a new one while keeping the old data at its existing location.

**Comment by ClearlyClaire** [comment]
One issue with migrating existing media files is that it will break references to their URL across the whole fediverse, meaning that when remote caches are cleaned (or when there isn't any remote cache), the media file would not be able to be fetched again.

**Comment by renchap** [comment]
Good point! I am not very familiar with Mastodon's code and the various protocols, is this only for media files (`MediaAttachment`) or for other files as well?
Do other instances update their locally-cached statuses from the original instance regularly (so they could fetch the new location), or is it permanently stored once the status is fetched and the media URL needs to never change?

My solution #3 (using Shrine) would allow to move off Paperclip while retaining the existing URLs for existing files. Would it be something you can consider?

Also, how urgent / needed is migrating away from Paperclip for the core team? This has been open since 2018 so I assume it is not urgent, but I am wondering if you are seeing any limitations from (old) Paperclip?

**Comment by katafrakt** [comment]
+1 for the idea of using Shrine. It's a great solution and seem to better support the migration case.

There are [some migration guides](https://gist.github.com/jmevans0211/b9abb28ed51b13a46df451e87c901767) for Paperclip -> ActiveStorage, but they include using actual models in a migration code, which is a dangerous antipattern and should only be used when you know precisely with which version of code the migration would be run. I don't think this is viable for an open source project, such a Mastodon. In worst case, some instances would stay on the old version of code, being unable to smoothly upgrade, which opens up security concerns.

**Comment by TheEssem** [comment]
Oh nice, yeah Shrine seems to be more viable to migrate to than ActiveStorage while also keeping the performance benefits. Seems like it can also be used for audio/video handling if I'm reading the docs correctly?

**Comment by renchap** [comment]
> Seems like it can also be used for audio/video handling if I'm reading the docs correctly?

Correct, Shrine can handle multiple versions of your files (called derivatives), and extract custom metadata as well, using various plugins. It can also handle dynamic derivatives, where the file (most often image) is resized on-demand when the request arrives, to avoid generating thumbnails. This can be delegated to an external process as well (like `imgproxy`) to avoid hitting Rails with those requests.

**Comment by tf-anguskong** [comment]
I have an additional point in moving towards ActiveStorage sooner rather than later as well, ActiveStorage has this method of functionality https://edgeapi.rubyonrails.org/classes/ActiveStorage/Blob.html#method-i-url that should be implemented ASAP. 

I think the biggest limitation of self-hosting is bandwidth (specifically upload) and with ActiveStorage's `service_url_for_direct_upload` method media uploads and the like would be able to be uploaded directly to S3 without going through the server first. Meaning a lot more people could host within the own homelab environments. 

I.E. I'm stuck hosting Mastodon on a cloud instance specifically because my home bandwidth (800 down/20 up) wouldn't be able to handle constant media uploads to S3. The S3 storage I have works great for serving the content now but I need the functionality to have the media immediately upload to it before I can move my instance on-premise. It's the ONLY thing keeping the instance in the cloud (Azure in my case)

**Comment by ShadowJonathan** [comment]
Paperclip's repo finally got marked as deprecated in july
